src/main.zig

```diff
@@ -345,6 +357,190 @@ const Node = union(enum) {
+                    }
+                }
+
+                // end of the line: insert a leaf with all present values
+                var ll = try newemptyll(stem, depth, alloc, crs);
+                node.children[stem[depth]] = Node{ .last_level = ll };
+                for (statediffs[statediff_index].suffix_diffs) |suffix_diff| {
+                    // will be left to `null` if suffix_dixx.current_value is `null`.
+                    if (suffix_diff.current_value) |sdiff| {
+                        ll.values[suffix_diff.suffix] = try alloc.create(Slot);
+                        std.mem.copy(u8, ll.values[suffix_diff.suffix].?, &sdiff);
+                    }
+                }
+
+                // "consume" a state diff
+                statediff_index += 1;
+            },
+        }
+    }
+
+    // Sanity check: ensure all commitments have been consumed.
+    if (commitment_index != commitments.len) {
+        return error.InvalidProofTooManyCommitments;
+    }
+
+    return root;
+}
+
+test "rebuild tree from proof" {
+    const json = std.json;
+    const state_diffs_json = @embedFile("./statediff_to_rebuild_from.json");
+    std.debug.print("state diff: {s}\n", .{state_diffs_json});
+
+    var statediffs = ArrayList(StemStateDiff).init(testing.allocator);
+    defer statediffs.deinit();
+    const T = []struct { stem: []u8, suffixDiffs: []struct { suffix: u8, currentValue: ?[]u8, newValue: ?[]u8 } };
+    var decoded = try json.parseFromSlice(T, testing.allocator, state_diffs_json, .{ .ignore_unknown_fields = true });
+    defer decoded.deinit();
+    for (decoded.value) |statediff| {
+        var new_stem_state_diff: StemStateDiff = StemStateDiff{ .stem = [_]u8{0} ** 31, .suffix_diffs = try testing.allocator.alloc(SuffixDiff, statediff.suffixDiffs.len) };
+        _ = try std.fmt.hexToBytes(new_stem_state_diff.stem[0..], statediff.stem[2..]);
+        for (statediff.suffixDiffs, 0..) |suffixdiff, i| {
+            new_stem_state_diff.suffix_diffs[i].suffix = suffixdiff.suffix;
+            if (suffixdiff.currentValue) |cv| {
+                new_stem_state_diff.suffix_diffs[i].current_value = [_]u8{0} ** 32;
+                _ = try std.fmt.hexToBytes(new_stem_state_diff.suffix_diffs[i].current_value.?[0..], cv[2..]);
+            } else {
+                // TODO check if I can use a default value instead, but that has some downstream implications that I need to master
+                new_stem_state_diff.suffix_diffs[i].current_value = null;
+            }
+            if (suffixdiff.newValue) |nv| {
+                new_stem_state_diff.suffix_diffs[i].new_value = [_]u8{0} ** 32;
+                _ = try std.fmt.hexToBytes(new_stem_state_diff.suffix_diffs[i].new_value.?[0..], nv[2..]);
+            } else {
+                new_stem_state_diff.suffix_diffs[i].new_value = null;
+            }
+        }
+        try statediffs.append(new_stem_state_diff);
+    }
```

Comment:
this code is all about decoding the statediff. It should be made into its own function.